tcmur_cmd_handler.c

```diff
@@ -459,6 +459,152 @@ static int handle_read(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+			       writecmd->iov_cnt, state->length, state->offset);
+}
+
+static void handle_format_unit_cbk(struct tcmu_device *dev,
+				   struct tcmulib_cmd *writecmd, int ret) {
+	struct tcmulib_cmd *origcmd = writecmd->cmdstate;
+	struct format_unit_state *state = origcmd->cmdstate;
+	uint8_t *sense = origcmd->sense_buf;
+	int rc;
+
+	writecmd->iovec->iov_base = state->write_buf;
+	state->offset += state->length;
+	state->done_blocks += state->length / dev->block_size;
+	if (state->done_blocks < dev->num_lbas)
+		dev->format_progress = (0x10000 * state->done_blocks) /
+					dev->num_lbas;
+
+	/* Check for last commmand */
+	if (state->done_blocks == dev->num_lbas) {
+		tcmu_dbg("last format cmd, done_blocks:%lu num_lbas:%lu block_size:%lu\n",
+			 state->done_blocks, dev->num_lbas, dev->block_size);
+		goto free_iovec;
+	}
+
+	if (state->done_blocks < dev->num_lbas) {
+		/* free iovec on every write, because seek I think in handlers
+		 * consume the iovec so can't resuse.
+		 */
+		free_iovec(writecmd);
+		if ((dev->num_lbas - state->done_blocks) * dev->block_size < state->length)
+			state->length = (dev->num_lbas - state->done_blocks) * dev->block_size;
+		if (alloc_iovec(writecmd, state->length)) {
+			ret = tcmu_set_sense_data(sense, HARDWARE_ERROR,
+						  ASC_INTERNAL_TARGET_FAILURE,
+						  NULL);
+			goto free_cmd;
+		}
+
+		/* copy incase handler changes it */
+		state->write_buf = writecmd->iovec->iov_base;
+
+		writecmd->done = handle_format_unit_cbk;
+
+		tcmu_dbg("next format cmd, done_blocks:%lu num_lbas:%lu block_size:%lu\n",
+			 state->done_blocks, dev->num_lbas, dev->block_size);
+		rc = async_handle_cmd(dev, writecmd, format_unit_work_fn);
+		if (rc != TCMU_ASYNC_HANDLED) {
+			tcmu_err(" async handle cmd failure");
+			ret = tcmu_set_sense_data(sense, MEDIUM_ERROR,
+						  ASC_WRITE_ERROR,
+						  NULL);
+			goto free_iovec;
+		}
+	}
+
+	return;
+
+free_iovec:
+	free_iovec(writecmd);
+free_cmd:
+	free(writecmd);
+	free(state);
+	pthread_mutex_lock(&dev->format_lock);
+	dev->flags &= ~FORMATTING;
+	pthread_mutex_unlock(&dev->format_lock);
+	aio_command_finish(dev, origcmd, ret);
+}
+
+static int handle_format_unit(struct tcmu_device *dev, struct tcmulib_cmd *cmd) {
+	struct tcmulib_cmd *writecmd;
+	struct format_unit_state *state;
+	size_t length = 1024 * 1024;
+	uint8_t *sense = cmd->sense_buf;
+	int ret;
+
+	pthread_mutex_lock(&dev->format_lock);
+	if (dev->flags & FORMATTING) {
+		pthread_mutex_unlock(&dev->format_lock);
+		return tcmu_set_sense_data(sense, NOT_READY,
+					  ASC_NOT_READY_FORMAT_IN_PROGRESS,
+					  &dev->format_progress);
+	}
+	dev->format_progress = 0;
+	dev->flags |= FORMATTING;
+	pthread_mutex_unlock(&dev->format_lock);
+
+	writecmd = calloc(1, sizeof(*writecmd));
+	if (!writecmd)
+		return SAM_STAT_TASK_SET_FULL;
+	writecmd->done = handle_format_unit_cbk;
+	writecmd->cmdstate = cmd;
+
+	state = calloc(1, sizeof(*state));
+	if (!state)
+		goto free_iov;
```

Comment:
I think the gotos and unraveling is in a odd order.

Here you would do goto free_cmd since we have not allocated the iov yet.